File: LeetCode/Python/Medium/1043-Partition-Array-for-Maximum-Sum/solution.py
'''
Solution 1: pure recursive solution. Note: this solution gets Time Limit Exceeded when run. DO NOT submit this one!
'''
class Solution:
    def maxSumAfterPartitioning(self, arr: List[int], k: int) -> int:
        result=0

        n=len(arr)

        index=0

        result=self.helper(arr, n, k, index)

        return result

    def helper(self, arr: List[int], n: int, k: int, index: int) -> int:
        if index >= n:
            return 0

        subproblem_solution=0

        subarray_max=0

        stop=min(index + k, n)

        for i in range(index, stop):
            subarray_max=max(subarray_max, arr[i])

            subarray_length=(i - index) + 1

            subarray_sum=subarray_max * subarray_length

            recursive_call_solution=subarray_sum + self.helper(arr, n, k, i + 1)

            subproblem_solution=max(subproblem_solution, recursive_call_solution)

        return subproblem_solution

'''
Solution 2: recursion + memoization.
This solution does pass the online judge.
'''
class Solution:
    def __init__(self):
        self.memo=list()

    def maxSumAfterPartitioning(self, arr: List[int], k: int) -> int:
        result=0

        n=len(arr)

        self.memo=[-1 for _ in range(0, n)]

        index=0

        result=self.helper(arr, n, k, index)

        return result

    def helper(self, arr: List[int], n: int, k: int, index: int) -> int:
        if index >= n:
            return 0

        if self.memo[index]!=-1:
            return self.memo[index]

        subproblem_solution=0

        subarray_max=0

        stop=min(index + k, n)

        for i in range(index, stop):
            subarray_max=max(subarray_max, arr[i])

            subarray_length=(i - index) + 1

            subarray_sum=subarray_max * subarray_length

            recursive_call_solution=subarray_sum + self.helper(arr, n, k, i + 1)

            subproblem_solution=max(subproblem_solution, recursive_call_solution)

        self.memo[index]=subproblem_solution

        return subproblem_solution

File: LeetCode/Rust/Medium/1043-Partition-Array-for-Maximum-Sum/solution.rs
impl Solution
{
    pub fn max_sum_after_partitioning(arr: Vec<i32>, k: i32) -> i32
    {
        let mut result: i32=0;

        let n: usize=arr.len();

        let mut memo: Vec<i32>=vec![-1; n];

        let mut index: usize=0;

        result=Self::helper(&arr, &mut memo, n, k, index);

        result
    }

    fn helper(arr: &Vec<i32>, memo: &mut Vec<i32>, n: usize, k: i32, index: usize) -> i32
    {
        if index >= n
        {
            return 0;
        }

        if memo[index]!=-1
        {
            return memo[index];
        }

        let mut subproblem_solution: i32=0;

        let stop: usize=usize::min(index + (k as usize), n);

        let mut subarray_max: i32=0;

        for i in index..stop
        {
            subarray_max=i32::max(subarray_max, arr[i]);

            let subarray_length: usize=(i - index) + 1;

            let subarray_total: i32=subarray_max * (subarray_length as i32);

            let recursive_call_solution: i32=subarray_total + Self::helper(arr, memo, n, k, i + 1);

            subproblem_solution=i32::max(subproblem_solution, recursive_call_solution);
        }

        memo[index]=subproblem_solution;

        subproblem_solution
    }
}